include/device.h

```diff
@@ -215,6 +347,78 @@ struct device {
 #endif
 };
 
+/**
+ * @brief Get the handle for a given device
+ *
+ * @param dev the device for which a handle is desired.
+ *
+ * @return the handle for the device.
+ */
+static inline device_handle_t
+device_get_handle(const struct device *dev)
+{
+	extern const struct device __device_start[];
+
+	/* TODO: If/when devices can be constructed that are not part of the
+	 * fixed we'll need another solution.
+	 */
+	return (device_handle_t)(dev - __device_start);
+}
+
+/**
+ * @brief Get the device corresponding to a handle.
+ *
+ * @param dev_handle the device handle
+ *
+ * @return the device that has that handle, or a null pointer if @p
+ * dev_handle does not identify a device.
+ */
+static inline const struct device *
+device_from_handle(device_handle_t dev_handle)
+{
+	extern const struct device __device_start[];
+	extern const struct device __device_end[];
+	const struct device *dev = NULL;
+	size_t numdev = __device_end - __device_start;
+
+	if (dev_handle < numdev) {
+		dev = &__device_start[dev_handle];
+	}
+
+	return dev;
+}
+
+/**
+ * @brief Get the set of devices on which a given device depends
+ *
+ * @param dev the device for which a dependencies are desired.
+ *
+ * @param count pointer to a place to store the number of devices provided at
+ * the returned pointer.  The value is not set if the call returns a null
+ * pointer.  The value may be set to zero.
+ *
+ * @return a pointer to a sequence of @p *count device handles, or a null
+ * pointer if @p dh does not provide dependency information.
+ */
+static inline const device_handle_t *
+device_get_requires_handles(const struct device *dev,
+			    size_t *count)
+{
+	const device_handle_t *rv = dev->handles;
+
+	if (rv != NULL) {
+		size_t i = 0;
+
+		while ((rv[i] != DEVICE_HANDLE_ENDS)
+		       && (rv[i] != DEVICE_HANDLE_SEP)) {
+			++i;
+		}
```

Comment:
@tbursztyka What you see above is how you extract the first set.   Does that answer your question in https://github.com/zephyrproject-rtos/zephyr/pull/29644#issuecomment-727814647 about how the sets are distinguished?